test/coreneuron/test_datareturn.py

```diff
@@ -172,56 +176,52 @@ def run(tstop, mode):
     print("CoreNEURON run")
     h.CVode().cache_efficient(1)
     coreneuron.enable = True
+    coreneuron.verbose = 0
     coreneuron.gpu = bool(
         distutils.util.strtobool(os.environ.get("CORENRN_ENABLE_GPU", "false"))
     )
-    cell_permute_values = (1, 2) if coreneuron.gpu else (0, 1)
 
-    coreneuron.cell_permute = cell_permute_values[0]
-    for mode in [0, 1, 2]:
-        run(tstop, mode)
-        tst = model.data()
-        max_unpermuted = h.Vector(std).sub(h.Vector(tst)).abs().max()
-        print("mode ", mode)
-        print("max diff permute=%d = %g" % (coreneuron.cell_permute, max_unpermuted))
-
-        coreneuron.cell_permute = cell_permute_values[1]
-        run(tstop, mode)
-        tst = model.data()
-        max_permuted = h.Vector(std).sub(h.Vector(tst)).abs().max()
-        print("max diff permute=%d = %g" % (coreneuron.cell_permute, max_permuted))
-
-        pc.nthread(2)
+    results = []
+    cell_permute_values = (1, 2) if coreneuron.gpu else (0, 1)
+    for mode, nthread, cell_permute in itertools.product(
+        [0, 1, 2], [1, 2], cell_permute_values
+    ):
+        pc.nthread(nthread)
+        coreneuron.cell_permute = cell_permute
         run(tstop, mode)
-
         tst = model.data()
-        max_permuted_thread = h.Vector(std).sub(h.Vector(tst)).abs().max()
-
-        coreneuron.enable = False
-
+        max_diff = (
+            h.Vector(v for k, v in std).sub(h.Vector(v for k, v in tst)).abs().max()
+        )
+        assert len(std) == len(tst)
         print(
-            "max diff permuted with %d threads = %g"
-            % (pc.nthread(), max_permuted_thread)
+            "mode={} nthread={} permute={} max diff={}".format(
+                mode, nthread, cell_permute, max_diff
+            )
         )
-
-        assert max_unpermuted < 1e-10
-        assert max_permuted < 1e-10
-        assert max_permuted_thread < 1e-10
-        pc.nthread(1)
+        # print([x for x in h.Vector(std).sub(h.Vector(tst))])
+        results.append(max_diff)
+        if max_diff > 1e-10:
+            for i in range(len(std)):
+                nrn_key, nrn_val = std[i]
+                tst_key, tst_val = tst[i]
+                assert nrn_key == tst_key
+                if abs(nrn_val - tst_val) >= 1e-10:
+                    print(i, nrn_key, nrn_val, tst_val, abs(nrn_val - tst_val))
+
+    assert all(max_diff < 1e-10 for max_diff in results)
 
     if __name__ != "__main__":
         # tear down
+        coreneuron.enable = False
```

Comment:
Just to say this was a bit of a guess; I assumed this "tear down" section was for `pytest` mode where `coreneuron` settings might otherwise pollute other tests, but I didn't verify this.

Reply:
That's right. But I guess the whole fragment may not be needed as it appears nrn/test/CMakeLists.txt runs this in isolation.

Reply:
tear-down / clean-up is not bad idea anyway